src/main/webapp/app/components/commentIcon/CommentIcon.tsx

```diff
@@ -0,0 +1,245 @@
+import { SizeProp } from '@fortawesome/fontawesome-svg-core';
+import { APP_DATETIME_FORMAT, CLOSE_TOOLTIP_DURATION_MILLISECONDS } from 'app/config/constants/constants';
+import { observer } from 'mobx-react';
+import { openCommentsIdStore, openCommentScrollPosition, commentsToDeleteStore } from 'app/stores/open-comment.store';
+import DefaultTooltip from 'app/shared/tooltip/DefaultTooltip';
+import { runInAction } from 'mobx';
+import { TextFormat } from 'react-jhipster';
+
+export interface ICommentIconProps {
+  id: string;
+  size?: SizeProp;
+  comments: Comment[];
+  onCreateComment: (content: string) => void;
+  onResolveComment: (index: number) => void;
+  onUnresolveComment: (index: number) => void;
+  onDeleteComments: (indices: number[]) => void;
+}
+
+const CommentIcon = observer((props: ICommentIconProps) => {
+  let color: string;
+  if (!props.comments || props.comments.length === 0) {
+    color = 'black';
+  } else if (props.comments.some(comment => comment.resolved === 'false')) {
+    color = 'gold';
+  } else {
+    color = 'green';
+  }
+
+  const timeoutId = useRef(null);
+
+  function handleMouseEnter() {
+    if (openCommentsIdStore.get() !== props.id) {
+      runInAction(() => openCommentsIdStore.set(props.id));
+    }
+    clearTimeout(timeoutId.current);
+  }
+
+  function handleMouseLeave() {
+    const id = setTimeout(() => {
+      if (props.id === openCommentsIdStore.get()) {
+        runInAction(() => openCommentsIdStore.set(null));
+      }
+    }, CLOSE_TOOLTIP_DURATION_MILLISECONDS);
+
+    timeoutId.current = id;
+  }
+
+  return (
+    <div onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
+      <DefaultTooltip
+        overlayStyle={{ opacity: 1 }}
+        overlayInnerStyle={{ minWidth: '400px', padding: 0, borderColor: 'secondary' }}
+        placement="left"
+        destroyTooltipOnHide
+        visible={props.id === openCommentsIdStore.get()}
+        overlay={
+          <CommentBox
+            openCommentId={props.id}
+            comments={props.comments}
+            onCreateComment={props.onCreateComment}
+            onResolveComment={props.onResolveComment}
+            onUnresolveComment={props.onUnresolveComment}
+            onDeleteComments={props.onDeleteComments}
+          />
+        }
+      >
+        <div>
+          <FontAwesomeIcon size={props.size || 'lg'} id={props.id} icon={faComment} color={color} />
+        </div>
+      </DefaultTooltip>
+    </div>
+  );
+});
+
+export interface ICommentBoxProps {
+  comments: Comment[];
+  openCommentId: string;
+  onCreateComment: (content: string) => void;
+  onResolveComment: (index: number) => void;
+  onUnresolveComment: (index: number) => void;
+  onDeleteComments: (indices: number[]) => void;
+}
+
+const CommentBox = observer((props: ICommentBoxProps) => {
+  const scrollContainerRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
```

Comment:
How come these useEffects have no dependency arrays?

Reply:
I want the effect to run on every rerender, not just when the component mounts. As for the second one, I only want it to run on unmount, so I think I can just add an empty dependency array.